topics api  new combined PR (#5585)
GoogleChrome/developer.chrome.com: site/en/docs/privacy-sandbox/topics/index.md

layout: 'layouts/doc-post.njk'
title: 'Topics API: developer guide'
subhead: >
  Learn how to work with the API, including how to use Chrome flags for testing.
description: >
  Learn how to work with the API, including how to use Chrome flags for testing.
date: 2022-01-25
updated: 2023-03-08
authors:
  - samdutton
---

## Implementation status
{% Partial 'privacy-sandbox/ps-implementation-status.njk' %}

## Try the Topics API

Topics is not currently available by default in any version of Chrome, but you can activate the API in two ways, as a single user or at scale:

-   The Topics API demo allows you to try it out as a single user.
-   The Topics origin trial allows you to try the API at scale with your website users.

### Try the demo

The demo of the Topics API is at [topics-demo.glitch.me](https://topics-demo.glitch.me/). It explains how to try out and debug the API for a single user.

You can also run the Topics [colab](/docs/privacy-sandbox/topics/colab/) to try out the Topics [classifier model](/docs/privacy-sandbox/topics/topic-classification/#classifier-model).

> ### Take part in a Topics origin trial

Reviewer: ```suggestion
### Test Topics in an origin trial
```

This may clarify that one can't find a "Topics origin trial" but rather has to find the relevence and measurement origin trial to test Topics.

```

{% Aside 'warning' %}
This snippet of code above is provided only to show how the Topics JavaScript API might be used. API design is subject to change.
{% endAside %}

### Use headers to access and observe topics

Rather than using the Topics JavaScript API from an iframe, topics can be accessed and marked as observed by using [request](https://developer.mozilla.org/docs/Web/API/Request/headers) and [response](https://developer.mozilla.org/docs/Web/API/Response/headers) headers:

-   Topics can be accessed from the `Sec-Browsing-Topics` header of a [`fetch()`](https://developer.mozilla.org/docs/Web/API/fetch) or [`XHR`](https://developer.mozilla.org/docs/Glossary/XHR_(XMLHttpRequest)) request.
-   Topics that were provided in a request header can be marked as observed by setting an `Observe-Browsing-Topics: ?1` header on the response to the request. The browser will then use those topics (that were included in the request header) for calculating topics of interest for a user.

> Using request and response headers to access topics and mark them as observed can be much more performant than using the JavaScript API from an iframe. For example, the header mechanism could be used when a `fetch()` request is made to an ad server. No iframe required! For more on this technique, check out the demo page.

Reviewer: ```suggestion
Using request and response headers to access topics and mark them as observed can be much more performant than using the JavaScript API from an iframe. For example, the header mechanism could be used when a `fetch()` request is made to an ad server. No iframe required! For more on this technique, check out the [demo](/docs/privacy-sandbox/topics/demo#the-topics-api-headers-demo).
```

The **Observed-by context domains (hashed)** column provides the hashed value of a hostname for which a topic was observed.

### View topics inferred for hostnames

You can also view the topics inferred by the Topics [classifier model](https://github.com/patcg-individual-drafts/topics#:~:text=classifier%20model) for one or more hostnames in `chrome://topics-internals`.

<figure>
{% Img src="image/80mq7dk16vVEg8BBhsVe42n6zn82/SOTuE2ljC55PaYll1UP1.png",
  alt="The chrome://topics-internals page with Classifier panel selected.",
  width="800", height="695" %}
  <figcaption>The chrome://topics-internals page Classifier panel shows topics selected, hosts visited, and model version and path.</figcaption>
</figure>

The current implementation of the Topics API infers topics from hostnames only; not from any other part of a URL.

Use hostnames only (without protocol or path) to view inferred topics from the `chrome://topics-internals` Classifier. `chrome://topics-internals` will display an error if you attempt to include a "/" in the Host field.

### View Topics API information

> Information is provided in `chrome://topics-internals` about the Topics API implementation and settings, such as the [taxonomy](https://github.com/jkarlin/topics/blob/main/taxonomy_v1.md) version and epoch duration. These values reflect default settings for the API or parameters successfully set [from the command line](#feature-flags). This is handy for checking that command-line flags have worked as expected. In the example below, `time_period_per_epoch` has been set to 15 seconds (the default is seven days).

Reviewer: ```suggestion
You can find information about the Topics API implementation and settings, such as the [taxonomy](https://github.com/jkarlin/topics/blob/main/taxonomy_v1.md) version and epoch duration, in `chrome://topics-internals`. These values reflect default settings for the API or parameters successfully set [from the command line](#feature-flags). This may be helpful to confirm that command line flags have worked as expected.

  In the example below, `time_period_per_epoch` has been set to 15 seconds (the default is seven days).
```
